wire-ios/Wire-iOS/Sources/UserInterface/Settings/Backup/BackupLocalStore/BackupLocalStore.swift

```diff
@@ -0,0 +1,170 @@
+    UserLocalStore: UserLocalStoreProtocol,
+    MessageLocalStore: MessageLocalStoreProtocol,
+    ConversationLocalStore: ConversationLocalStoreProtocol {
+
+    /// The context to call `perform(schedule:_:)` on if needed.
+    private let context: NSManagedObjectContext
+
+    private let userLocalStore: UserLocalStore
+    private let conversationLocalStore: ConversationLocalStore
+    private let messageLocalStore: MessageLocalStore
+
+    func countModels() async throws -> (userCount: Int, conversationCount: Int, messageCount: Int) {
+        let userCount = try await userLocalStore.totalUserCountForBackup()
+        let conversationCount = try await conversationLocalStore.totalConversationCountForBackup()
+        let messageCount = try await messageLocalStore.totalMessageCountForBackup()
+        return (userCount, conversationCount, messageCount)
+    }
+
+    // MARK: -
+
+    func fetchAllUsers() -> AsyncThrowingStream<UserBackupModel, any Error> {
+        AsyncThrowingStream { continuation in
+            Task<Void, Never> {
+                do {
+                    let users = try await userLocalStore.fetchAllUsersForBackup()
+                    await context.perform {
+                        for user in users {
+                            autoreleasepool {
+                                if let backupUser = UserBackupModel(user) {
+                                    continuation.yield(backupUser)
+                                }
+                            }
+                        }
+                    }
+                    continuation.finish()
+                } catch {
+                    continuation.finish(throwing: error)
+                }
+            }
+        }
+    }
+
+    // MARK: -
+
+    func fetchAllConversations() -> AsyncThrowingStream<ConversationBackupModel, any Error> {
+        AsyncThrowingStream { continuation in
+            Task<Void, Never> {
+                do {
+                    let conversations = try await conversationLocalStore.fetchAllConversationsForBackup()
+                    await context.perform {
+                        for conversation in conversations {
+                            autoreleasepool {
+                                if let backupConversation = ConversationBackupModel(conversation) {
+                                    continuation.yield(backupConversation)
+                                }
+                            }
+                        }
+                    }
+                    continuation.finish()
+                } catch {
+                    continuation.finish(throwing: error)
+                }
+            }
+        }
+    }
+
+    // MARK: -
+
+    func fetchAllMessages() -> AsyncThrowingStream<MessageBackupModel, any Error> {
+        AsyncThrowingStream { continuation in
+            Task<Void, Never> {
+                do {
+                    let messages = try await messageLocalStore.fetchAllMessagesForBackup()
```

Comment:
issue: the fetch request inside this method has a batch limit, so `messages` is actually a batched array. We then take this array and iterate through all elements and push them through the stream. I'm wondering if by doing this we eliminate the benefits of the batching since those elements may all accumulate in memory. What do you think?

Reply:
The batching happens under the hood: it looks like an array of all fetched messages, but in reality only the first batch is initially loaded and the next batch is loaded lazily when that array index is accessed. Certainly we don't load all elements at once, but I guess we do load them all in to memory as we go through the array, I'm just wondering if there's some clean up we need to do to ensure the memory consumption reasonable.

In any case, it would be good to try to profile it to get an idea of what's going on. Also, in case we do get a memory warning, what would happen?